### PR TITLE
Alternative List API

### PR DESCRIPTION
**Pull request by robertbastian**

### CODE REVIEW

File: experimental/list_formatter/src/list_formatter.rs [hunk 1]
```diff
@@ -2,47 +2,52 @@
...
-            .load_resource(&DataRequest {
-                options: locale.into().into(),
-                metadata: Default::default(),
-            })?
-            .take_payload()?;
-        Ok(Self { data, style })
-    }
+macro_rules! constructor {
+    ($name: ident, $marker: ty, $doc: literal) => {
+        #[doc = concat!("Creates a new [`ListFormatter`] that produces a ", $doc, "-type list. See the [CLDR spec]",
```
Comment (sffc) [inline experimental/list_formatter/src/list_formatter.rs:21]: Praise: This `#[doc = concat!()]` trick does actually appear to work.  This is a nice option when needing to auto-generate doc strings in macros.

http://icu4x-pr-artifacts.storage.googleapis.com/gha/811a0b003fbdd211572f01b5e44d7a834355d92d/docs/icu_list/struct.ListFormatter.html

File: experimental/list_formatter/src/lib.rs [hunk 2]
```diff
@@ -49,6 +49,9 @@
 extern crate alloc;
 
 mod list_formatter;
+#[cfg(feature = "provider_transform_internals")]
+pub mod provider;
+#[cfg(not(feature = "provider_transform_internals"))]
```
Comment (sffc) [inline experimental/list_formatter/src/lib.rs:54]: Issue: Although I would like `mod provider` to be private, it is used in public trait bounds.  The compiler usually complains about this, but I've seen cases where it doesn't detect a private type used in a public signature.  I think the macro might be obscuring this fact.  CC @Manishearth

**Review by sffc — DISMISSED** [review]
LGTM though I would like @Manishearth to weigh in on the question about public/private types.

**Review by Manishearth — DISMISSED** [review]

File: experimental/list_formatter/src/provider.rs [hunk 3]
```diff
@@ -12,9 +12,7 @@ use alloc::borrow::Cow;
 use icu_provider::yoke::{self, *};
 use writeable::{LengthHint, Writeable};
 
-/// Symbols and metadata required for [`ListFormatter`](crate::ListFormatter). Requires
-/// feature "provider_transform_internals".
-// (the feature makes it pub)
+/// Symbols and metadata required for [`ListFormatter`](crate::ListFormatter).
 #[icu_provider::data_struct(
     ErasedListV1Marker,
```
Comment (sffc) [inline experimental/list_formatter/src/provider.rs:17]: Nit: make this private by declaring it manually.

**Review by Manishearth — APPROVED** [review]

**Review by Manishearth — APPROVED** [review]

**Pull request merged**